resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java

```diff
@@ -0,0 +1,191 @@
+        List<DnsCacheEntry> oldEntries = resolveCache.get(hostname);
+        final List<DnsCacheEntry> entries;
+        if (oldEntries == null) {
+            List<DnsCacheEntry> newEntries = new ArrayList<DnsCacheEntry>(8);
+            oldEntries = resolveCache.putIfAbsent(hostname, newEntries);
+            entries = oldEntries != null? oldEntries : newEntries;
+        } else {
+            entries = oldEntries;
+        }
+        return entries;
+    }
+
+    @Override
+    public void cache(String hostname, InetAddress address, long originalTtl, EventLoop loop) {
+        if (maxTtl == 0) {
+            return;
+        }
+
+        final int ttl = Math.max(minTtl, (int) Math.min(maxTtl, originalTtl));
+        final List<DnsCacheEntry> entries = cachedEntries(hostname);
+        final DnsCacheEntry e = new DnsCacheEntry(hostname, address);
+
+        synchronized (entries) {
+            if (!entries.isEmpty()) {
+                final DnsCacheEntry firstEntry = entries.get(0);
+                if (firstEntry.cause() != null) {
+                    assert entries.size() == 1;
+                    firstEntry.cancelExpiration();
+                    entries.clear();
+                }
+            }
+            entries.add(e);
+        }
+
+        scheduleCacheExpiration(entries, e, ttl, loop);
+    }
+
+    @Override
+    public void cache(String hostname, Throwable cause, EventLoop loop) {
+        if (negativeTtl == 0) {
+            return;
+        }
+
+        final List<DnsCacheEntry> entries = cachedEntries(hostname);
+        final DnsCacheEntry e = new DnsCacheEntry(hostname, cause);
+
+        synchronized (entries) {
+            final int numEntries = entries.size();
+            for (int i = 0; i < numEntries; i ++) {
+                entries.get(i).cancelExpiration();
+            }
+            entries.clear();
+            entries.add(e);
+        }
+
+        scheduleCacheExpiration(entries, e, negativeTtl, loop);
+    }
+
+    private static void cancelExpiration(List<DnsCacheEntry> entries) {
+        final int numEntries = entries.size();
+        for (int i = 0; i < numEntries; i++) {
+            entries.get(i).cancelExpiration();
+        }
+    }
+
+    private void scheduleCacheExpiration(final List<DnsCacheEntry> entries,
+                                         final DnsCacheEntry e,
+                                         int ttl,
+                                         EventLoop loop) {
+        e.scheduleExpiration(loop, new OneTimeTask() {
+                    @Override
+                    public void run() {
+                        synchronized (entries) {
+                            entries.remove(e);
+                            if (entries.isEmpty()) {
+                                resolveCache.remove(e.hostname());
+                            }
+                        }
+                    }
+                }, ttl, TimeUnit.SECONDS);
+    }
+}
```

Comment:
Could you also add `toString()`? e.g. `DefaultDnsCache(minTtl: ?, maxTtl: ?, negativeTtl: ?, <some stats>)`